core/offchain/src/api.rs

```diff
@@ -158,6 +162,33 @@ impl<Storage, KeyProvider> OffchainExt for Api<Storage, KeyProvider> where
 		Ok(CryptoKeyId(id))
 	}
 
+	fn authority_pubkey(&self, kind: CryptoKind) -> Result<Vec<u8>, ()> {
+		match self.read_key(None, kind) {
+			Ok(key) => {
+				let public = match key {
+					Key::Sr25519(pair) => pair.public().encode(),
+					Key::Ed25519(pair) => pair.public().encode(),
+				};
+
+				Ok(public)
+			},
+			Err(e) => Err(e),
+		}
+	}
+
+	fn network_state(&self) -> Result<OpaqueNetworkState, ()> {
+		let external_addresses = self.network_state
+			.external_addresses()
+			.into_iter()
```

Comment:
why do we `into_iter()` and `collect()`? Maybe better return `Vec` from `NetworkStateInfo` or have associated type there?

Reply:
It's a `HashSet` in `core/network`, but I transform to a `Vec`, since `Encode/Decode` is only implemented for `Vec` currently and the support for `Vec` in WASM is really good. I liked your suggestion of returning `impl Iterator<Item = Multiaddr>`, but as written in the reply for that suggestion that's sadly not possible with Rust in traits currently.

Do you see a shorter way of getting around this allocation? The only alternative I see would be to implement `Encode/Decode` for `HashSet` as well and see if/how well it's supported in WASM.

Reply:
Well you can always have an associated type and collect to that, but it might be a bit cumbersome to handle.

But maybe we can just change `HashMap` to `Vec`, since it seems we already get an iterator from `Swarm` so you'd just collect into a `Vec` on that level. I guess collecting to a `HashSet` allows you to deduplicate, but is that needed?

Reply:
Yes, good point. I talked to Pierre and the related code just uses a `HashSet` so that it gets serialized into a JavaScript object as opposed to a Map. So we can collect into a `Vec`.